cpp/src/arrow/array/array_primitive.h

```diff
@@ -119,7 +129,7 @@ class NumericArray : public PrimitiveArray {
   IteratorType end() const { return IteratorType(*this, length()); }
 
  protected:
```

Comment:
The use of `protected` is odd here; nothing inherits from NumericArray inside arrow. Is it intended to support user subclassing of NumericArray? Or could we just mark NumericArray final? @pitrou

Reply:
This might keep same style from `PrimitiveArray`, which initializes `raw_values_` in ctor. Don't know should we change here